### PR TITLE
Add zalgo-promise

### PR DESCRIPTION
**Pull request by daniel-shuy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/zalgo-promise/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// Type definitions for zalgo-promise 1.0
+// Project: https://github.com/krakenjs/zalgo-promise#readme
+// Definitions by: Daniel Shuy <https://github.com/daniel-shuy>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export { ZalgoPromise } from './promise';
```
Comment (peterblazejewicz) [inline types/zalgo-promise/index.d.ts:6]: I think it has to support UMD, so Global script usage:
https://github.com/krakenjs/zalgo-promise#global
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-a-umd-library

Comment (daniel-shuy) [reply types/zalgo-promise/index.d.ts:6]: Oh right, thanks!

What about CommonJS? Is there a way to export for both CommonJS and ES (as a class module)?
```ts
// UMD
export as namespace ZalgoPromise;

// CommonJS
export = ZalgoPromise;    // fails with "An export assignment cannot be used in a module with other exported elements"

// ES
export class ZalgoPromise<R> {
```

Comment (daniel-shuy) [reply types/zalgo-promise/index.d.ts:6]: Nevermind, turns out the `README` is outdated. The CommonJS and UMD modules are imported as:
```html
<!-- UMD -->
<script src="zalgo-promise.js"></script>
<script>
    new ZalgoPromise.ZalgoPromise( ... );
</script>
```js
// CommonJS
var ZalgoPromise = require('zalgo-promise');

new ZalgoPromise.ZalgoPromise( ... );
```
This removes the need to exported it as a class module, therefore there is no issue exporting for UMD, CommonJS and ES.

I've updated the PR accordingly, thanks!

File: types/zalgo-promise/promise.d.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    static onPossiblyUnhandledException(handler: (err: any) => void): {cancel: () => void};
+
+    // to support conditional promising returning method
+    static try<X, A extends readonly any[]>(method: (...args: A) => ZalgoPromise<X> | undefined, context?: any, args?: Partial<A>): ZalgoPromise<X | undefined>;
+    static try<X, A extends readonly any[]>(method: (...args: A) => ZalgoPromise<X>, context?: any, args?: Partial<A>): ZalgoPromise<X>;
+    static try<Y, A extends readonly any[]>(method: (...args: A) => Y, context?: any, args?: Partial<A>): ZalgoPromise<Y>;
+
+    static delay(delay: number): ZalgoPromise<void>;
+
+    static isPromise(value: any): boolean;
```
Comment (peterblazejewicz) [inline types/zalgo-promise/promise.d.ts:57]:
```suggestion
    static isPromise(value: any): value is Promise<any>;
```
should this be type predicate?

Comment (daniel-shuy) [reply types/zalgo-promise/promise.d.ts:57]: Looking at the implementation (https://github.com/krakenjs/zalgo-promise/blob/v1.0.48/src/promise.js#L455, https://github.com/krakenjs/zalgo-promise/blob/v1.0.48/src/utils.js), it is not just doing a simple typecheck. It is not only returning true if `value` is `ZalgoPromise`/`Promise`, but also if it is an object with a `then` function.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@daniel-shuy thx!

**Comment by daniel-shuy** [comment]
Thanks @peterblazejewicz!

**Pull request merged**